Facilitator readme file (#244)
mosaicdao/facilitator: README.md

# Facilitator 

Mosaic facilitator is an executable which enables atomic token transfers across two blockchains. Facilitator executable facilitates stake & mint and Redeem & unstake of any EIP20 token using mosaic gateways. 

## Prerequisite
1. Docker version `18.09.1` or above.
2. Docker compose version `1.23.2` or above.
3. Node version `11` or above.
4. Mosaic chains executable. 
   ```

>    npm i @openst/mosaic-chains -g

Author: Simply `mosaic start ropsten`

>    npm i @openst/mosaic-chains -g

Author: yes

   
      
      web3.eth.sendTransaction(
      {
         from:<funder_address>, 
          to: <origin_worker>,
          value:<Fund_in_wei
      })
      
   *Gas on auxiliary*: Facilitator also needs to pay for transaction fee on auxiliary chain. For testnet [mosaic faucet](https://github.com/mosaicdao/faucet) can be used to get fund. Alternatively, OST on value chain can also be converted to base token on auxiliary chain in order to pay for gas.
   
   *Value token for bounty*: Facilitator needs to stake bounty to gateway/co-gateway in order to perform stake & mint and redeem & unstake. Bounty token on the origin chain for existing mosaic chain is simple token and on the auxiliary chain bounty is base token. 
   
   On testnet, bounty can be funded to workers using [mosaic faucet](https://github.com/mosaicdao/faucet).
   
**7. White list workers**: origin and auxiliary workers should be whitelisted in stakepool and redeem pool contracts respectively. Below commands will whitelist the workers. 

   1. Clone git repository:
   ```

>        git clone https://github.com/mosaicdao/mosaic-chains.git

Reviewer: we asked the user to install `mosaic-chains` globally above. Have them clone it again here might be a problem

Author: It's for the whitelist-worker script. Any other idea?

 
 Documentation of deployment of stake and redeem pool can be found [here](https://github.com/mosaicdao/mosaic-chains#stake-pool).
 
 **3. Deploy subgraphs**: Origin and auxiliary subgraphs are needed for facilitator executable. Below commands will deploy origin and auxiliary subgraph. 

*Origin subgraph*: Below command will deploy origin subgraph. 
```
mosaic subgraph <origin-chain-identifier> <auxiliary-chain-identifier> origin <origin-graph-rpc-admin-url> <origin-graph-ipfs-url> 

```

*Auxiliary subgraph*: Below command will deploy auxiliary sugraph.

```
mosaic subgraph <origin-chain-identifier> <auxiliary-chain-identifier> auxiliary <aux-graph-rpc-admin-url> <aux-graph-ipfs-url> 

```

`<graph-rpc-admin-url>` and `<graph-ipfs-url>` was displayed while starting origin and auxiliary chain.

By default subgraph command will deploy subgraph for OST gateways, optionally it also accepts `--mosaic-config`, `--gateway-config` and `--gateway-address` option to deploy subgraph for other gateways. 

Mosaic config i.e `mosaic.json` can be found in [here](https://github.com/mosaicdao/mosaic-chains/tree/develop/chains) inside `<origin-chain>` folder.

> Gateway config i.e. `<gatewayaddress>.json` can be found in [here](https://github.com/mosaicdao/mosaic-chains/tree/develop/chains) inside `<origin-chain>/<auxiliary-chain>/<gateway-address.json`.

Reviewer 2: ```suggestion
Gateway config i.e. `<gatewayaddress>.json` can be found [here](https://github.com/mosaicdao/mosaic-chains/tree/develop/chains) inside `<origin-chain>/<auxiliary-chain>/<gateway-address.json`.
```
Same comment as above.

More documentation about `subgraph` command can be found [here](https://github.com/mosaicdao/mosaic-chains#subgraph-deployment).

> **4. Facilitator init**: Facilitator init command initialize the facilitator which creates database, worker addresses and create facilitator config file. Facilitator config file is needed to start facilitator.  

Reviewer 2: ```suggestion
**4. Facilitator init**: Facilitator init command initializes the facilitator. The command creates database, worker addresses and create facilitator config file. Facilitator config file is mandatory to start facilitator.  
```

```
facilitator init --mosaic-config <mosaic-config> --aux-chain-id <aux-chain-id> --origin-password <origin-password> --auxiliary-password <auxiliary-password> --origin-rpc <origin-rpc> --auxiliary-rpc <auxiliary-rpc> --origin-graph-ws <origin-graph-ws> --origin-graph-rpc <origin-graph-rpc> --auxiliary-graph-ws <auxiliary-graph-ws> --auxiliary-graph-rpc <auxiliary-graph-rpc> --db-path <db-path> --force

```

> * Replace `<mosaic-config>` with location where mosaic config is present.

Reviewer 2: ```suggestion
* Replace `<mosaic-config>` with file location where mosaic config is present.
```

* Replace `<aux-chain-id>` with auxiliary chain id. 
* Replace `<origin-password>` with the password required to encrypt the worker account of origin chain created with this command. It will be required to unlock worker account while starting facilitator.
* Replace `<auxiliary-password>` with the password required to encrypt the worker account of auxiliary chain created with this command. It will be required to unlock worker account while starting facilitator.
* Replace `<origin-rpc>` with origin chain's rpc url.
* Replace `<auxiliary-rpc>` with auxiliary chain's rpc url.
* Replace `<origin-graph-ws>` with origin ws graph endpoint.
* Replace `<origin-graph-rpc>` with origin rpc graph admin rpc endpoint.
* Replace `<auxiliary-graph-ws>` with auxiliary ws graph endpoint.
* Replace `<auxiliary-graph-rpc>` with auxiliary rpc graph admin endpoint.
* Replace `<db-path>` with the database path. It is the path for `sqlite` database. If not provided,it would create it.
* `--force` option is used to forcefully override facilitator config. It is optional parameter.

> Facilitator init can be also done with `--gateway-config <gateway-config>` option. Gateway config i.e. `<gatewayaddress>.json` can be found in [here](https://github.com/mosaicdao/mosaic-chains/tree/develop/chains) inside `<origin-chain>/<auxiliary-chain>/<gateway-address.json`.

Reviewer 2: ```suggestion
Facilitator init can also be done with `--gateway-config <gateway-config>` option. Gateway config i.e. `<gatewayaddress>.json` can be found [here](https://github.com/mosaicdao/mosaic-chains/tree/develop/chains). It's inside `<origin-chain>/<auxiliary-chain>/<gateway-address.json`.
```
Same comment as above.

> Replace <gateway-config> with location where gateway config is present.

Reviewer 2: ```suggestion
Replace <gateway-config> with file location where gateway config is present.
```

This command will generate facilitator config file which is needed to start facilitator. 

**5. Set ENV variables**: Facilitator init commands create worker keys which are encrypted with password. Enviornment variables need to be set to unlock the encrypted keystore file. 

Set below env variables after replacing address and password.
```
export MOSAIC_ADDRESS_PASSW_<Address>=<origin-password>
export MOSAIC_ADDRESS_PASSW_<Address>=<auxiliary-password>
```
Above variables will also be produced with the output of `facilitator init` command.

**6. Fund facilitator workers for gas and bounty**: Workers created in facilitator init step needs to be funded for gas and bounty on both chains.

   *Gas on origin*: Facilitator needs to pay for gas on origin chain. Origin worker address created in `facilitator init` step must be funded to pay for transaction fee. 
   

>    There are various ways to fund on origin chain, for testnet any public faucet can be used. Below web3 transaction can also be done, if there is fund in an existing key.

Reviewer 2: ```suggestion
   There are various ways to fund worker on origin chain. For testnet any public faucet can be used. Below web3 transaction can be done, if there is fund in an existing key.
```

   
      
      web3.eth.sendTransaction(
      {
         from:<funder_address>, 
          to: <origin_worker>,
          value:<Fund_in_wei
      })
      

>    *Gas on auxiliary*: Facilitator also needs to pay for transaction fee on auxiliary chain. For testnet [mosaic faucet](https://github.com/mosaicdao/faucet) can be used to get fund. Alternatively, OST on value chain can also be converted to base token on auxiliary chain in order to pay for gas. This is explained in detail in section [request stake]().

Reviewer 2: [request stake]() can be hyperlinked.

   
   *Value token for bounty*: Facilitator needs to stake bounty to gateway/co-gateway in order to perform stake & mint and redeem & unstake. Bounty token on the origin chain for existing mosaic chain is simple token and on the auxiliary chain bounty is base token. 
   
   On testnet, bounty can be funded to workers using [mosaic faucet](https://github.com/mosaicdao/faucet).
   

> **7. White list workers**: origin and auxiliary workers should be whitelisted in stakepool and redeem pool contract respectively. Below commands will whitelist the workers. 

Reviewer 2: ```suggestion
**7. White list workers**: origin and auxiliary workers should be whitelisted in stakepool and redeem pool contracts respectively. Below commands will whitelist the workers. 
```

   1. Clone git repository:
   ```
       git clone https://github.com/mosaicdao/mosaic-chains.git
   ```
   2. Set below enviornment variables
   

>     

Reviewer 2: why two new lines?

    export ORIGIN_WEB3_ENDPOINT='replace_with_origin_web3_endpoint';
    export AUXILIARY_WEB3_ENDPOINT='replace_with_auxiliary_web3_endpoint';
    export AUXILIARY_CHAIN_ID='replace_with_auxiliary_chain_id';
    export MOSAIC_CONFIG_PATH='replace_with_mosaic_config_path';
    export ORIGIN_WORKER_ADDRESS='replace_with_origin_worker_address';
    export AUXILIARY_WORKER_ADDRESS='replace_with_auxiliary_worker_address';
    export ORIGIN_WORKER_EXPIRATION_HEIGHT='replace_with_origin_expiration_height';
    export AUXILIARY_WORKER_EXPIRATION_HEIGHT='replace_with_auxiliary_expiration_height';

origin and auxiliary worker addresses are generated with `facilitator init` step. 
Mosaic config path for supported chain should be available on `~/.mosaic/<origin-chain>/mosaic.json` where `<origin-chain>` is origin chain identifier e.g. `ropsten`.  

origin and auxiliary worker expiration height is block height from current block for which worker keys are whitelisted. If current block is 1000 and expiration height is set to 100 then worker keys will be whitelisted for 1100 block.

   3. run command:
     

>      npm run whitelist-workers

Reviewer 2: will he run this from facilitator repo?

Author: No, step 1 is about cloning mosaic-chains.

    * Replace `<gateway-config>` with the path to gateway-config.json.	    
    
  * **Note** : Both `--mosaic-config` and `--gateway-config` are together not allowed in command.

    

> 

Reviewer: We haven't added a section in here to help users to just move tokens around using facilitators hosted by others.
